src/Futhark/Analysis/Interference.hs

```diff
@@ -0,0 +1,333 @@
+
+analyseBody ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  Body KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseBody lumap inuse body = analyseStms lumap inuse $ bodyStms body
+
+analyseStms ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  Stms KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseStms lumap inuse0 stms = do
+  inScopeOf stms $ do
+    foldM
+      ( \(inuse, lus, graph) stm -> do
+          (inuse', lus', graph') <- analyseStm lumap inuse stm
+          return (inuse', lus' <> lus, graph' <> graph)
+      )
+      (inuse0, mempty, mempty)
+      $ stmsToList stms
+
+analyseSegOp ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  SegOp lvl KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseSegOp lumap inuse (SegMap _ _ _ body) =
+  analyseKernelBody lumap inuse body
+analyseSegOp lumap inuse (SegRed _ _ binops _ body) =
+  segWithBinOps lumap inuse binops body
+analyseSegOp lumap inuse (SegScan _ _ binops _ body) = do
+  segWithBinOps lumap inuse binops body
+analyseSegOp lumap inuse (SegHist _ _ histops _ body) = do
+  (inuse', lus', graph) <- analyseKernelBody lumap inuse body
+  (inuse'', lus'', graph') <- mconcat <$> mapM (analyseHistOp lumap inuse') histops
+  return (inuse'', lus' <> lus'', graph <> graph')
+
+segWithBinOps ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  [SegBinOp KernelsMem] ->
+  KernelBody KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+segWithBinOps lumap inuse binops body = do
+  (inuse', lus', graph) <- analyseKernelBody lumap inuse body
+  (inuse'', lus'', graph') <-
+    mconcat
+      <$> mapM
+        (analyseSegBinOp lumap inuse')
+        binops
+  return (inuse'', lus' <> lus'', graph <> graph')
+
+analyseSegBinOp ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  SegBinOp KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseSegBinOp lumap inuse (SegBinOp _ lambda _ _) =
+  analyseLambda lumap inuse lambda
+
+analyseHistOp ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  HistOp KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseHistOp lumap inuse histop =
+  analyseLambda lumap inuse (histOp histop)
+
+analyseLambda ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  InUse ->
+  Lambda KernelsMem ->
+  m (InUse, LastUsed, Graph VName)
+analyseLambda lumap inuse (Lambda _ body _) =
+  analyseBody lumap inuse body
+
+-- | Perform interference analysis on the given statements. The result is a
+-- triple of the names currently in use, names that hit their last use somewhere
+-- within, and the resulting graph.
+analyseKernels ::
+  LocalScope KernelsMem m =>
+  LastUseMap ->
+  Stms KernelsMem ->
+  m (Graph VName)
+analyseKernels lumap stms = do
+  (_, _, graph) <- analyseKernels' lumap stms
+  -- Now, we need to insert edges between memory blocks which differ in size, if
+  -- they are in DefaultSpace.
```

Comment:
Why is DefaultSpace special?  Does this imply that we treat DefaultSpace memory blocks with different sizes as overlapping/interfering?  Why?

Reply:
The commit (and my associated blog post) has more information: d482f9d3c5eb988ce2c3f5f44a4ed74600bff92d

Reply:
Add a note in the code itself.